client/src/App.tsx

```diff
@@ -1,67 +1,262 @@
 
+  useEffect(() => {
+    if (!isTokenVisible) return;
+
+    const timeoutId = setTimeout(() => {
+      setIsTokenVisible(false);
+    }, 30000);
+
+    return () => clearTimeout(timeoutId);
+  }, [isTokenVisible]);
+
   const getData = async () => {
     const response = await fetch(API_URL);
-    const { data } = await response.json();
+    const data = await response.json();
     setData(data);
   };
 
   const updateData = async () => {
+    const keyPair = ec.keyFromPrivate(accessToken, 'hex');
+
+    const signature = keyPair.sign(data).toDER('hex');
+
     await fetch(API_URL, {
-      method: "POST",
-      body: JSON.stringify({ data }),
+      method: 'POST',
+      body: JSON.stringify({ data, signature }),
       headers: {
-        Accept: "application/json",
-        "Content-Type": "application/json",
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
       },
     });
 
     await getData();
   };
 
   const verifyData = async () => {
-    throw new Error("Not implemented");
+    console.log('verifyData');
+    try {
+      const data = await fetch(`${API_URL}/verify`, { method: 'GET' });
+      const { data: resData } = await data.json();
+
+      console.log(resData[0].data, resData[0].signature, resData[0].key);
+
+      const keyPair = ec.keyFromPrivate(accessToken, 'hex');
+
+      console.log(keyPair);
+      const isVerified = ec.verify(
+        resData[0].data,
+        resData[0].signature,
+        resData[0].key,
+        'hex'
+      );
+
+      console.log(isVerified);
+
+      setTampered(!isVerified);
+    } catch (error) {
+      setTampered(true);
+      console.log(error);
+    }
+  };
+
+  const tamperDataToCheck = async () => {
+    const keyPair = ec.keyFromPrivate(accessToken, 'hex');
+
+    const signature = keyPair.sign(data).toDER('hex');
+
+    await fetch(API_URL, {
+      method: 'POST',
+      body: JSON.stringify({
+        data: data + ' tampered',
+        signature: signature + 'tampered',
+      }),
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+    });
+
+    await getData();
+  };
+
+  const getAccessToken = async () => {
+    try {
+      const response = await fetch(`${API_URL}/access-token`, {
+        method: 'POST',
+        body: JSON.stringify({ data: {} }),
+        headers: {
+          Accept: 'application/json',
+          'Content-Type': 'application/json',
+        },
+      });
+      const { data } = await response.json();
+
+      setIsTokenVisible(true);
+      setIsTokenBtnVisible(false);
+
+      console.log({ data });
+
+      setAccessToken(data.accessToken);
+
+      return accessToken;
+    } catch (error) {
+      console.log(error);
+      return null;
+    }
+  };
+
+  const recoverData = async () => {
+    console.log('recoverData');
+
+    const resData = await fetch(`${API_URL}/history`, { method: 'GET' });
+    const { data: history } = await resData.json();
```

Comment:
What if the history has been compromised too, all of the history has been removed?

Reply:
Well to ensure that the history can't be compromised is, we can make history immutable. One other thing that can be implemented is to notify the client for every update or pre-update on data and history, so that quick actions can be taken.

To answer your question, we can do regular back of historical data to an external and secure location. If the primary database is compromised or records are deleted, a recent and accurate backup can be used to restore the historical data.